node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1289,11 +1295,196 @@ async fn fetch_block_worker(
                 // want to retry.
                 error!(
                     ?err,
-                    ?block_hash,
+                    ?block_header_hash,
                     "failed to download block with infinite retries"
                 );
             }
         }
+
+        if let Some(finality_signatures) =
+            fetch_finality_signatures_by_block_header(block_header, ctx).await?
+        {
+            ctx.effect_builder
+                .put_signatures_to_storage(finality_signatures)
+                .await;
+        };
+    }
+}
+
+struct BlockSignaturesCollector(Option<BlockSignatures>);
+
+impl BlockSignaturesCollector {
+    fn new() -> Self {
+        Self(None)
+    }
+
+    fn add_signatures(&mut self, maybe_signatures: Option<BlockSignatures>) {
+        let signatures = match maybe_signatures {
+            Some(sigs) => sigs,
+            None => return,
+        };
+
+        match &mut self.0 {
+            None => {
+                self.0 = Some(signatures);
+            }
+            Some(old_sigs) => {
+                for (pub_key, signature) in signatures.proofs {
+                    old_sigs.insert_proof(pub_key, signature);
+                }
+            }
+        }
+    }
+
+    fn check_if_sufficient(
+        &self,
+        validator_weights: &BTreeMap<PublicKey, U512>,
+        finality_threshold_fraction: Ratio<u64>,
+    ) -> bool {
```

Comment:
Maybe we should have just one method that tries to add the signature and:
* validate it
* if correct add
* if sufficient informs the caller

Reply:
See here: https://github.com/casper-network/casper-node/pull/3125#discussion_r905099348

That approach will give us a function named similarly to `validate_and_add_if_correct_and_check_if_sufficient`. I prefer to keep the validation as a separate step.

Reply:
I still think we should have it be part of the `add` method as we could end up storing invalid signatures due to developer's mistake (if he forgets to verify them before adding) but it's your decision.

Reply:
I added fc200a04b7cb092a1e2cbdabd9cffbd2f6060c0a which introduces such function.